### DISCUSSION TITLE
Google API in Java

### DISCUSSION THREAD

**Discussion by vighneshb04** [community/community#136935] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

General

### Body

To add and use Google APIs in a Java project, you'll generally need to follow these steps:

1. Set Up a Google Cloud Project and Enable APIs
Create a Google Cloud Project: If you haven't already, create a project in the Google Cloud Console.
Enable the API: Go to the API Library in the Google Cloud Console, search for the specific API you need (e.g., Google Sheets API, Google Drive API), and enable it.
Create Credentials: After enabling the API, create credentials (OAuth 2.0 client IDs, API keys, or service accounts) depending on your use case.
2. Add Google API Client Libraries to Your Java Project
You can add the necessary dependencies to your project using a build tool like Maven or Gradle.

For Maven:
Add the required dependencies in your pom.xml file. For example, if you want to use the Google Sheets API:

xml
Copy code
<dependencies>
    
    <dependency>
        <groupId>com.google.api-client</groupId>
        <artifactId>google-api-client</artifactId>
        <version>1.33.0</version>
    </dependency>

    
    <dependency>
        <groupId>com.google.oauth-client</groupId>
        <artifactId>google-oauth-client-jetty</artifactId>
        <version>1.33.1</version>
    </dependency>

    
    <dependency>
        <groupId>com.google.apis</groupId>
        <artifactId>google-api-services-sheets</artifactId>
        <version>v4-rev614-1.25.0</version>
    </dependency>
</dependencies>
For Gradle:
Add the following dependencies to your build.gradle file:

gradle
Copy code
dependencies {
    // Google API Client
    implementation 'com.google.api-client:google-api-client:1.33.0'

    // Google OAuth Client
    implementation 'com.google.oauth-client:google-oauth-client-jetty:1.33.1'

    // Google Sheets API
    implementation 'com.google.apis:google-api-services-sheets:v4-rev614-1.25.0'
}
3. Use the Google API in Your Java Code
Here's an example using the Google Sheets API:

java
Copy code
import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.SheetsScopes;
import com.google.api.services.sheets.v4.model.ValueRange;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.security.GeneralSecurityException;
import java.util.Collections;
import java.util.List;

public class GoogleSheetsExample {
    private static final String APPLICATION_NAME = "Google Sheets API Java Quickstart";
    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
    private static final List<String> SCOPES = Collections.singletonList(SheetsScopes.SPREADSHEETS_READONLY);
    private static final String CREDENTIALS_FILE_PATH = "/credentials.json"; // Replace with your credentials file path

    private static Credential getCredentials() throws IOException {
        InputStream in = GoogleSheetsExample.class.getResourceAsStream(CREDENTIALS_FILE_PATH);
        GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(JSON_FACTORY, new InputStreamReader(in));
        GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(
                GoogleNetHttpTransport.newTrustedTransport(), JSON_FACTORY, clientSecrets, SCOPES)
                .setDataStoreFactory(new FileDataStoreFactory(new java.io.File("tokens")))
                .setAccessType("offline")
                .build();
        return new AuthorizationCodeInstalledApp(flow, new LocalServerReceiver()).authorize("user");
    }

    public static void main(String[] args) throws IOException, GeneralSecurityException {
        final NetHttpTransport HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
        Sheets service = new Sheets.Builder(HTTP_TRANSPORT, JSON_FACTORY, getCredentials())
                .setApplicationName(APPLICATION_NAME)
                .build();

        String spreadsheetId = "your-spreadsheet-id"; // Replace with your Google Sheets ID
        String range = "Sheet1!A1:D5"; // Replace with the range you want to read
        ValueRange response = service.spreadsheets().values().get(spreadsheetId, range).execute();
        List<List<Object>> values = response.getValues();

        if (values == null || values.isEmpty()) {
            System.out.println("No data found.");
        } else {
            for (List row : values) {
                System.out.printf("%s, %s\n", row.get(0), row.get(1)); // Print columns A and B
            }
        }
    }
}
4. Run the Java Program
Ensure that the credentials.json file is placed in your project's resource directory.
Run the Java program. On the first run, it will prompt you to authenticate with Google. After authentication, it will store your credentials locally for future use.
Notes:
Replace "your-spreadsheet-id" with the actual ID of your Google Sheet.
Customize the code as needed depending on the specific Google API you're using.
This process can be adapted for other Google APIs by changing the dependencies and modifying the API-specific code.

**Comment by vighneshb04** [comment]
done

**Comment by ebndev** [comment, marked as answer]
Hi @vighneshb04, thanks for being a part of the GitHub Community!

While we appreciate you wanting to contribute to Discussions, please note: We made the decision to disable the ability to earn Achievements in our Community in order to discourage users from participating in coordinated or inauthentic activity like rapid questions and answers in order to earn badges. You can learn more about this decision in our announcement post here [Achievements will no longer be available in the Community](https://github.com/orgs/community/discussions/106536).

Note that GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this, and other related posts. 

Any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
